pymatviz/ptable.py

```diff
@@ -703,293 +752,210 @@ def histogram(
 
 
 def ptable_heatmap(
-    values: ElemValues,
+    data: pd.DataFrame | pd.Series | dict[str, list[list[float]]],
     *,
-    log: bool | Normalize = False,
-    ax: plt.Axes | None = None,
-    count_mode: ElemCountMode = ElemCountMode.composition,
+    # Heatmap specific
+    colormap: str = "viridis",
+    # heat_mode: Literal["value", "fraction", "percent"] = "value",
+    # infty_color: str = "lightskyblue",
+    # na_color: str = "white",
+    log: bool = False,
+    # Figure-scope
+    # f_block_voffset: float = 0.5,
+    on_empty: Literal["hide", "show"] = "hide",
+    hide_f_block: bool | Literal["AUTO"] = "AUTO",
+    plot_kwargs: dict[str, Any] | None = None,
+    # Axis-scope
+    # tile_size: float | tuple[float, float] = 0.9,
+    ax_kwargs: dict[str, Any] | None = None,
+    # Symbol
+    # fmt: str | Callable[..., str] | None = None,
+    symbol_text: str | Callable[[Element], str] = lambda elem: elem.symbol,
+    symbol_pos: tuple[float, float] | None = None,
+    symbol_color: str = "black",  # TODO: control this by kwargs?
+    symbol_kwargs: dict[str, Any] | None = None,
+    # Values
+    show_values: bool = True,
+    values_pos: tuple[float, float] | None = None,
+    values_color: str = "black",  # TODO: control this by kwargs?
+    values_kwargs: dict[str, Any] | None = None,
+    # Colorbar
+    show_cbar: bool = True,
     cbar_title: str = "Element Count",
-    cbar_range: tuple[float | None, float | None] | None = None,
+    # cbar_range: tuple[float | None, float | None] | None = None,
     cbar_coords: tuple[float, float, float, float] = (0.18, 0.8, 0.42, 0.05),
+    cbar_title_kwargs: dict[str, Any] | None = None,
     cbar_kwargs: dict[str, Any] | None = None,
-    colorscale: str = "viridis",
-    show_scale: bool = True,
-    show_values: bool = True,
-    infty_color: str = "lightskyblue",
-    na_color: str = "white",
```

Comment:
Need some discussion on NaN and infinity colors here, as the new implementation would replace them. Do we want to treat them specially or keep consistent with the new plotter?

Meanwhile I think passing NaN or inf should be the edge case for heatmap plotter?

Reply:
for `ptable_heatmap_ratio` the objective is definitely to show 0 and inf as anomalous values. if you're looking at the ratio of two element distributions 0 means an element is not in the numerator, whereas inf means it's not in the denominator which is worth highlighting with separate colors. if you're not doing a ratio plot, 0 is not important to highlight and inf should very rarely happen but still worth having a special color for inf imo to cover edge cases and point out to the users there's something anomalous in the heatmap values

Reply:
Thanks for the input. Yes it's always good to have the flexibility to assign a special color to a special value (inf/NaN/zero in this case). 

I need to have a good think about how to actually achieve this :) Perhaps we need another preprocessor for heatmap plotter (which should be pretty easy, because it's just a `{element: single value}` pair). I would take care of this.

Meanwhile, while I'm working on this, I noticed my original plan (migrate everything at the same time) doesn't seem to work because there are simply too many functionalities to rewrite. I would instead refactor with the base function, and add one functionality at a time (see the top of the PR) :)

And here is what we got right now:
![ptable_test](https://github.com/janosh/pymatviz/assets/80093591/399f61bb-3aad-42eb-b8a9-72b77662ac95)

Reply:
looks good, thanks for sharing! are there any visual differences compared to before? maybe post the same plot generated from `main` for comparison?

Reply:
btw, thanks for working on this! good timing as i was planning to make some changes to `ptable_heatmap_plotly` this weekend which might have caused merge conflicts here (though probably not) so i'll hold off on that just in case

Reply:
> looks good, thanks for sharing! 

Thanks!

> are there any visual differences compared to before? maybe post the same plot generated from `main` for comparison?

I haven't done this comparison yet. Would do tomorrow :)

Reply:
Here is the heatmap from the plotter in main:

There are a few minor differences:
- The font looks like bold by default, and in "white" color of course.
- The color of the values would switch between black and white automatically depending on the background color, that's a good feature to have. I would migrate this.
- The default decimal places are a bit different, but still should be an easy fix.

![ptable_test](https://github.com/janosh/pymatviz/assets/80093591/5b63cc04-f662-40a4-ac10-3669e519328a)